build.gradle

```diff
@@ -19,40 +19,78 @@ ext {
 
-    // This file is used as a dummy output in place of the graalvm repository directory.
-    // This prevents buildGraal from re-triggering cloneAndPatchRepository.
+    // These files are used as a dummies in place of the graalvm repository directory.
+    // This prevents buildGraal from re-triggering clone/patch operations and breaking incremental builds.
     graalCloneDummy = "$tempDir/graal_cloned_time"
+    graalPatchDummy = "$tempDir/graal_patched_time"
+    graalBuildDummy = "$tempDir/graal_build_dummy"
 }
 
-subprojects {
-    apply plugin: 'maven-publish'  // Plugin to publish artifacts to a maven repo
-    apply plugin: 'signing'
-    apply plugin: 'java'
+// Why is this necessary? Who knows!
+tasks.register('clean', Delete) {
+    delete rootProject.buildDir
+}
 
-    task cloneAndPatchGraal(type: Exec) {
+task cloneGraal(type: Exec) {
 
-        // Task details
-        group = 'Other'
-        description = 'Clones a specific version of Graal and patches the source code with the required changes for Conclave.'
+    // Task details
+    group = 'Other'
+    description = 'Clone a specific version of Graal. Delete existing graal clone if present.'
 
-        def patchingScript = "${rootProject.projectDir}/scripts/cloneAndPatch.sh"
-        def graalPatch = "${rootProject.projectDir}/patches/graal.patch"
+    def cloneScript = "${rootProject.projectDir}/scripts/cloneGraal.sh"
 
-        inputs.files(patchingScript, graalPatch)
-        outputs.files(graalCloneDummy)
+    // Re-run this only if the version or commit ID have changed.
+    inputs.property('graalVersion', graal_version)
+    inputs.property('graalCommitID', graal_commit_id)
+    inputs.file(cloneScript)
 
-        doFirst {
-            mkdir tempDir
-            workingDir tempDir
-            commandLine "$patchingScript", "$graal_version", "$graal_commit_id"
-            //commandLine "ls"
+    // Used to detect if subsequent tasks should re-run
+    outputs.file(graalCloneDummy)
 
-            // Write current date to graal clone dummy file. If this task is run again, the content of this file
-            // will change and subsequent tasks will run.
-            writeDateToFile(graalCloneDummy)
-        }
+    doFirst {
+        mkdir tempDir
+        workingDir tempDir
+        commandLine "$cloneScript", "$graal_version", "$graal_commit_id"
+    }
+
+    doLast {
+        // Write current date-time to dummy file. If this task is run again, the content of this file
+        // will change and subsequent tasks will re-run.
+        writeDateToFile(graalCloneDummy)
+    }
+}
+
+task patchGraal(type: Exec) {
+    dependsOn(cloneGraal)
+
+    // Task details
+    group = 'Other'
+    description = 'Patch the graal repository with Conclave specific changes.'
+
+    def patchScript = "${rootProject.projectDir}/scripts/patchGraal.sh"
+    def patchFile = "${rootProject.projectDir}/patches/graal.patch"
+
+    inputs.files(patchScript, patchFile, graalCloneDummy)
```

Comment:
Can you leave a comment explaining why we cannot monitor the actual files that are used as input for this task?